sdb/commands/zfs/dbuf.py

```diff
@@ -0,0 +1,123 @@
+        if dd.dd_parent:
+            pname = DslDir.name(dd.dd_parent) + '/'
+        return pname + dd.dd_myname.string_().decode("utf-8")
+
+
+class Dataset():
+
+    @staticmethod
+    def name(ds):
+        if not ds:
+            return 'MOS'
+        name = DslDir.name(ds.ds_dir)
+        if not ds.ds_prev:
+            sn = ds.ds_snapname.string_().decode("utf-8")
+            if len(sn) == 0:
+                sn = '%UNKNOWN_SNAP_NAME%'
+            name += '@' + sn
+        return name
+
+
+class Objset():
+
+    @staticmethod
+    def name(os):
+        return Dataset.name(os.os_dsl_dataset)
+
+
+class Dbuf(sdb.Locator, sdb.PrettyPrinter):
+
+    names = ["dbuf"]
+    input_type = "dmu_buf_impl_t *"
+    output_type = "dmu_buf_impl_t *"
+
+    def _init_argparse(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            '-o',
+            '--object',
+            type=int,
+            help='filter: only dbufs of this object')
+        parser.add_argument(
+            '-l',
+            '--level',
+            type=int,
+            help='filter: only dbufs of this level')
+        parser.add_argument(
+            '-b',
+            '--blkid',
+            type=int,
+            help='filter: only dbufs of this blkid')
+        parser.add_argument(
+            '-d',
+            '--dataset',
+            type=str,
+            help='filter: only dbufs of this dataset name (or "MOS")')
+        parser.add_argument(
+            '-H',
+            '--has-holds',
+            action='store_true',
+            help='filter: only dbufs that have nonzero holds')
+
+    def pretty_print(self, dbufs):
+        print("{:>20} {:>8} {:>4} {:>8} {:>5} {}".format(
+            "addr", "object", "lvl", "blkid", "holds", "os"))
+        for dbuf in dbufs:
+            print("{:>20} {:>8d} {:>4d} {:>8d} {:>5d} {}".format(
+                hex(dbuf),
+                int(dbuf.db.db_object),
+                int(dbuf.db_level),
+                int(dbuf.db_blkid),
+                int(dbuf.db_holds.rc_count),
+                Objset.name(dbuf.db_objset)))
+
+    def filter(self, db):
+        if self.args.object and db.db.db_object != self.args.object:
+            return False
+        if self.args.level and db.db_level != self.args.level:
+            return False
+        if self.args.blkid and db.db_blkid != self.args.blkid:
+            return False
+        if self.args.has_holds and db.db_holds.rc_count == 0:
+            return False
+        if self.args.dataset and Objset.name(db.db_objset) != self.args.dataset:
+            return False
+        return True
+
+    def no_input(self) -> Iterable[drgn.Object]:
+        hash_map = self.prog["dbuf_hash_table"].address_of_()
+        for i in range(hash_map.hash_table_mask):
+            dbuf = hash_map.hash_table[i]
+            while dbuf:
+                if self.filter(dbuf):
+                    yield dbuf
+                dbuf = dbuf.db_hash_next
```

Comment:
would it be better to leverage `map` here? e.g.
```
+    @staticmethod
+    def yield_dbufs(dbuf) -> Iterable[drgn.Object]:
+        while dbuf:
+            yield dbuf
+            dbuf = dbuf.db_hash_next

     def no_input(self) -> Iterable[drgn.Object]:
         hash_map = self.prog["dbuf_hash_table"].address_of_()
         for i in range(hash_map.hash_table_mask):
-            dbuf = hash_map.hash_table[i]
-            while dbuf:
-                if self.filter(dbuf):
-                    yield dbuf
-                dbuf = dbuf.db_hash_next
+            yield from map(self.filter, Dbuf.yield_dbufs(hash_map.hash_table[i]))
```
With something like this, we could potentially move the `yield_dbufs` function into some "library", along with your other helper functions.

I'm don't feel strongly; I just wanted to bring up an alternate way to effectively accomplish the same thing. In this scheme, we'd also have to update `filter` to return `db` instead of `True`, and `None` instead of `False`.

Reply:
Is there a magic value like `None` that `filter` would return to indicate "this maps to nothing"?  I couldn't find any info about that in the docs https://docs.python.org/3/library/functions.html#map

Reply:
Looking more at the docs, I think the `filter` builtin might be what we want to use, instead of `map`; https://docs.python.org/3/library/functions.html#filter

I'm new here, so I could be wrong, but I do think using `map`, `filter`, etc. is more idomatic python, than the `while` loop. With that said, I'm OK leaving as-is; I don't feel strongly either way.